6_nodejs/week2/app.js

```diff
@@ -0,0 +1,110 @@
+import express from "express";
+import fs from "fs";
+const data = fs.readFileSync("./documents.json");
+let documents = JSON.parse(data);
+//console.log(documents);
+const app = express();
+const port = process.env.PORT || 3000;
+// Support parsing JSON requests
+app.use(express.json());
+// For parsing application/x-www-form-urlencoded
+//app.use(express.urlencoded({ extended: true }));
+app.get("/", (req, res) => {
+  res.send("This is a search engine.");
+});
+//---------------------------------------------------------------
+//GET /search
+/* This endpoint will accept a query parameter called q, short for query. 
+If q is not provided, the endpoint should return all documents.
+If q is provided, the endpoint should return the documents with some field that matches the value of q.
+Example response if we call GET   /search?q=hello :
+[
+    {
+        "id": 2,
+        "type": "doc",
+        "value": "hello world"
+    }
+]
+*/
+app.get("/search", (req, res) => {
+  const query = req.query.q; // Extract the 'q' query parameter from the request URL
+  if (!query) {
+    res.json(documents); // If 'q' is not provided, it returns all documents
+  } else {
+    // filter documents' objects, as doc, based on whether any field matches the query string
+    const filteredDocuments = documents.filter((doc) =>
+      Object.values(doc).some(
+        (value) => typeof value === "string" && value.includes(query)
+      )
+    );
+    if (filteredDocuments) {
+      res.json(filteredDocuments);
+    } else {
+      res.send("no match found");
```

Comment:
you might want to keep the endpoint response consistent - so if for a certain query, the response is a JSON with an array of objects, then if there is nothing found for a different query the endpoint should return a JSON with an empty array, not a string instead of JSON. That way you keep the type of data that the endpoint returns the same - it will always be a JSON and that will be easier to handle for anyone who is using your API.